detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/BracesOnIfStatements.kt

```diff
@@ -0,0 +1,178 @@
+ * else c
+ *
+ * // singleLine = 'always'
+ * if (a) { b } else { c }
+ * if (a) { b } else if (c) { d }
+ * // multiLine = 'always'
+ * if (a) { b }
+ * else { c }
+ * if (a) { b }
+ * else if (c) { d }
+ *
+ * // singleLine = 'consistent'
+ * if (a) { b } else { c }
+ * if (a) b else { c; d } // multi-expression
+ * // multiLine = 'consistent'
+ * if (a) { b }
+ * else { c }
+ * if (a) b
+ * else c
+ * </compliant>
+ */
+class BracesOnIfStatements(config: Config = Config.empty) : Rule(config) {
+
+    override val issue = Issue(
+        javaClass.simpleName,
+        Severity.Style,
+        "Braces do not comply with the specified policy",
+        Debt.FIVE_MINS
+    )
+
+    @Configuration("single-line braces policy")
+    private val singleLine: BracePolicy by config(BRACES_NEVER) { BracePolicy.getValue(it) }
+
+    @Configuration("multi-line braces policy")
+    private val multiLine: BracePolicy by config(BRACES_ALWAYS) { BracePolicy.getValue(it) }
+
+    override fun visitIfExpression(expression: KtIfExpression) {
+        super.visitIfExpression(expression)
+        if (expression.parent.parent is KtIfExpression) return
+
+        val branches: List<KtExpression> = walk(expression)
+        validate(branches, policy(expression))
+    }
+
+    private fun walk(expression: KtExpression): List<KtExpression> {
+        val list = mutableListOf<KtExpression>()
+        var current: KtExpression? = expression
+        while (current is KtIfExpression) {
+            current.then?.let { list.add(it) }
+            // Don't add `if` because it's an `else if` which we treat as one unit.
+            current.`else`?.takeIf { it !is KtIfExpression }?.let { list.add(it) }
+            current = current.`else`
+        }
+        return list
+    }
+
+    private fun validate(list: List<KtExpression>, policy: BracePolicy) {
+        when (policy) {
+            BracePolicy.Always -> {
+                if (!list.all { hasBraces(it) }) {
+                    val violator = list.first { !hasBraces(it) }
+                    report(violator, policy)
+                }
+            }
+
+            BracePolicy.Never -> {
+                if (!list.all { !hasBraces(it) }) {
+                    val violator = list.first { hasBraces(it) }
+                    report(violator, policy)
+                }
+            }
+
+            BracePolicy.Consistent -> {
+                val reference = hasBraces(list.first())
+                if (!list.all { hasBraces(it) == reference }) {
+                    val violator = list.first { hasBraces(it) != reference }
+                    report(violator, policy)
+                }
+            }
+        }
+    }
+
+    private fun report(violator: KtExpression, policy: BracePolicy) {
+        val iff = violator.parent.parent as KtIfExpression
+        val reported = when {
+            iff.then === violator -> iff.ifKeyword
+            iff.`else` === violator && isMultiStatement(violator) -> iff.ifKeyword
```

Comment:
This feels strange, would this work instead (I think this expresses more `else if`, and it localizes the report more targeted on where the problem is)
```suggestion
            iff.`else` === violator && violator is KtIfExpression -> violator.ifKeyword
```
Test case (not executed, just thought-validated):
`if (true) { println() } else if (false) println()` + always or consistent config. Original highlight on `if` at character 1, and expected highlight would be the second `if`.

Reply:
I would say the expected highlights are (based on the discussion we had yesterday about context):
```kotlin
// always:
if (true) { println() } else if (false) println()
                             ^^

// always:
if (a) b else { println(); println() }
^^

// never:
if (a) b else { println(); println() }
         ^^^^
```

I think at this point in the logic, if you need to check multi-statement or KtBlock, that means you're making an assumption on the structure or policy.

> It would be the second if, because we compare the first element against all others.

This confirms my assumption theory, I think we were talking about different policies, I meant for most of them, and you're talking about consistent.

This actually highlights a problem, and thanks for that. For consistent it makes sense to highlight the first if. That way we're not opinionated on flagging a missing/extra braces. The problem is that we would need to traverse up to find it. As the first option in the `when` we could add something like:
```kotlin
policy == Consistent ->
  // Walk up to the first `if` in an arbitrary long else-if chain.
  generateSequence(violator) { it.parent.parent as? KtIfExpression }.last()
```
and for other cases in the when we just highlight whatever the violator's branch was.